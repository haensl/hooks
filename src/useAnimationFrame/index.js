import { useRef, useEffect, useCallback } from 'react';

const useAnimationFrame = (callback) => {
  const requestRef = useRef();
  const previousTimeRef = useRef();
  const animate = useCallback((time) => {
    if (typeof previousTimeRef.current !== 'undefined') {
      const deltaTime = time - previousTimeRef.current;
      callback(deltaTime);
    }

    previousTimeRef.current = time;
    requestRef.current = requestAnimationFrame(animate);
  }, [previousTimeRef, requestRef, callback]);

  useEffect(() => {
    requestRef.current = requestAnimationFrame(animate);
    return () => {
      cancelAnimationFrame(requestRef.current);
    };
  }, [animate]);
};

export default useAnimationFrame;
