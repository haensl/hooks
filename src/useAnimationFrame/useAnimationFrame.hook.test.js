import { act, renderHook } from '@testing-library/react-hooks';
import '@testing-library/jest-dom';

import useAnimationFrame from './';

describe('useAnimationFrame', () => {
  let hasAnimated;
  let requestAnimationFrameSpy;

  beforeAll(() => {
    jest.useFakeTimers();
    hasAnimated = false;
    requestAnimationFrameSpy = jest.spyOn(
      window,
      'requestAnimationFrame'
    ).mockImplementation((callback) => {
      if (!hasAnimated) {
        setTimeout(callback(Date.now() + 10));
        hasAnimated = true;
      }
    });
  });

  afterAll(() => {
    jest.useRealTimers();
    requestAnimationFrameSpy.mockRestore();
  });

  beforeEach(() => {
    hasAnimated = false;
  });

  describe('when animating', () => {
    it('applies the animation', () => {
      const handler = jest.fn();
      renderHook(() => useAnimationFrame(handler));

      act(() => {
        jest.runOnlyPendingTimers();
      });

      expect(handler).toHaveBeenCalled();
    });
  });
});

