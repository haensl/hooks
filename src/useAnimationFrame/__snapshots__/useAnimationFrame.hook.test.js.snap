// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useAnimationFrame renders as expected 1`] = `
<span>
  useAnimationFrame test
</span>
`;
